default/python/universe_generation/empires.py

```diff
@@ -95,26 +85,101 @@ def min_planets_in_vicinity_limit(num_systems):
+        with the best minimum system merit of all such lists picked randomly from the ''systems_pool''.
+
+        Algorithm:
+        Make several attempts to find systems that match the condition
+        of being at least min_jumps apart.
+        Use the minimum merit of the best num_home_system systems found
+        as the merit of the current best set of systems.
+        On each attempt use the merit of the current best set of home
+        systems to truncate the pool of candidates.
+        """
+
+        # precalculate the system merits
+        for system in systems_pool:
+            if system not in self.system_merit:
+                self.system_merit[system] = calculate_home_system_merit(system)
+
+        # In the list of systems reverse sorted by merit, if the first
+        # num_home_systems are all min_jumps apart, then the merit of the
+        # num_home_systems nth system is the best_case_merit_lower_bound that
+        # can be achieved.
+        all_merit_system = sorted([(self.system_merit[s], s)
+                                   for s in systems_pool], reverse=True)
+        current_merit_lower_bound = 0
+
+        best_candidate = []
+
+        # cap the number of attempts to the smaller of the number of systems in the pool, or 100
+        attempts = min(100, len(systems_pool))
+
+        # Cap the number of attempts when the found number of systems is much less than the target
+        # num_home_systems because this indicates that the min_jumps is too large and/or the
+        # systems_pool is too small to ever succeed.
+        num_complete_misses_remaining = 4
+
+        while attempts and num_complete_misses_remaining:
+            # use a local pool of all candidate systems better than the worst threshold merit
+            all_merit_system = [(m, s) for (m, s) in all_merit_system if m > current_merit_lower_bound]
+            local_pool = {s for (m, s) in all_merit_system}
+
+            if len(local_pool) < self.num_home_systems:
+                if len(best_candidate) < self.num_home_systems:
+                    print ("Failing in find_home_systems_for_min_jump_distance because "
+                           "current_merit_lower_bound = {} trims local pool to {} systems "
+                           "which is less than num_home_systems {}.".format(
+                               current_merit_lower_bound, len(local_pool), self.num_home_systems))
+                break
+
+            attempts = min(attempts - 1, len(local_pool))
+
+            candidate = []
+            while local_pool:
+                member = random.choice(list(local_pool))
+                candidate.append(member)
+
+                # remove all neighbors from the local pool
+                local_pool -= set(get_systems_within_jumps(member, min_jumps))
+
+            # Count complete misses when number of candidates is not close to the target.
+            if len(candidate) < (self.num_home_systems - 4):
```

Comment:
Ok, here the variance on the length of candidates being within a range of about 4 you mentioned comes in, correct? So we have a complete miss when we lack more than 4 systems to meet the `num_home_system` goal.

Question: Isn't that variance dependend on the number of home systems we try to find? I'd expect that variance to be greater if I'm looking for 40 than 4 home systems. While it probably doesn't make sense to come up with complex calculations to reduce the complete miss treshold when looking for fewer systems (I guess the performance gain with a reduced treshold for e.g. 10 home systems will be negligible), we need to account for the case that the current hard limit of 40 players might be increased in the future.

We've been raising those hard limits in the past, so I expect that to happen again in the future. Then the hardcoded treshold of 4 might become to narrow, and I expect no one is going to remember that there is some hard coded treshold somewhere deep in the universe generation code that needs according adjustment.

So better make it scale properly with the number of home systems requested.